transcom/mymove
# B-20012 INT - GBLOC Switching for HQ User

## [B-20012](https://www13.v1host.com/USTRANSCOM38/assetdetail.v1?number=B-20012)

## Summary

Gives the HQ user the ability to switch GBLOCs so they can view the queues from all GBLOCs.

Adds a nullable query parameter to all queue endpoints that can specify a desired GBLOC to view as. If the requesting user is not an HQ user the parameter is ignored. The front end should never use this parameter for any non HQ user but I chose to ignore it in the back end to ensure access controls are still applied and to help prevent FE errors if somehow a non HQ user makes a request that includes this new parameter. 

Since the component that allows the user to select a GBLOC and the TableQueue that initiates the request are far apart in the component tree React Context was used to share state across distant components. [These docs](https://react.dev/learn/passing-data-deeply-with-context) and [this SO answer](https://stackoverflow.com/a/72758141) explains more about the approach used.

## Verification Steps for the Author

These are to be checked by the author.


## Verification Steps for Reviewers

These are to be checked by a reviewer.


### Setup to Run the Code

- [Instructions for starting storybook](https://transcom.github.io/mymove-docs/docs/frontend/setup/storybook)
- [Instructions for starting the MilMove application](https://transcom.github.io/mymove-docs/docs/getting-started/application-setup/)
- [Instructions for running tests](https://transcom.github.io/mymove-docs/docs/getting-started/development/testing)

### How to test
This affects a lot of places so this testing should be a lot of clicking around, and maybe trying to break it, to check nothing goes wrong.

1. Ensure the HQ Role FF is enabled.
2. Access the office app.
3. Login as an HQ user in any GBLOC.
4. In the header where the user's GBLOC is usually listed verify you have access to a drop down.
5. Select various GBLOCs from that dropdown and tab through the various queues ensuring everything displays correctly and updates when you select a new GBLOC or apply a filter or sort.
6. Spot check a couple selected GBLOC and selected Queue combinations with a non HQ user that is created in that GBLOC.
7. In Swagger, attempt to request each of the queues: Counseling (normal and PPM use the same endpoint), Moves, and Payment Requests while logged in as their usual user type (NOT HQ) with the viewAsGBLOC parameter set to a GBLOC other than that of the user you're logged in as. Verify the return value is for the logged in user's GBLOC rather than the requested GBLOC.


### Frontend


### Backend


### Database

#### Any new migrations/schema changes:


## Screenshots
<img width="596" alt="Screenshot 2024-06-28 at 10 48 27 AM" src="https://github.com/transcom/mymove/assets/158500142/e32630e0-1a99-41e7-96c5-38c184f573df">
<img width="1192" alt="Screenshot 2024-06-28 at 10 49 08 AM" src="https://github.com/transcom/mymove/assets/158500142/0df994c4-cb61-4390-9163-1670496b0b69">
